docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+The state format would change slightly to account for linking together the root cid of an ipfs file dag and the cluster-dag pinning its leaves
+
+```
+"<cid>": {
+  "name": <pinname>,
+  "cid": <cid>,
+  "clusterdag": <cidcluster>,
+  "allocations" : []
+}
+"<shard1-cid>": {
+  "name": "<pinname>-shard1",
+  "cid": <shard1-cid>,
+  "clusterdag": nil,
+  "allocations": [cluster peers]
+}
+....
+```
+
+Under this implementation replication of shards proceeds like replication of an ordinary pinned cid.  As a consequence shards can be pinned with a replication factor.
+
+### User APIs
+
+#### Importing files for sharding
+Data sharding must begin with a user facing endpoint.  Ideally cluster could stream files, or other data to be imported, across its HTTP API.  This way `ipfs-cluster-ctl` could launch the command for adding and sharding data in the usual way, and users also have the option of hitting the HTTP endpoint over the network.  Multiple files are typically packaged in an HTTP request as the parts of a `Content-Type: multi-part` request and HTTP achieves streaming with `Transfer-Type: chunked`.  Because golang http and mime/multipart [support chunked (a.k.a. streamed) transfer of multipart messages](https://gist.github.com/ZenGround0/49e4a1aa126736f966a1dfdcb84abdae), building the user endpoint into the HTTP API should be relatively straightforward.
+
+The golang multipart-reader exposes a reader to each part of the body in succession.  When the part's body is read completely, a reader to the next part's body can be generated by the multipart-reader.  The part meta-data (e.g. this part is a directory, that part is a file) and the data read from the part, must then be passed into the importer component to begin the next phase of the pipeline.  We can explore different methods of this transfer in pursuit of efficiency and simplicity.  One approach would be polling the part readers in the HTTP API and transferring data as it arrives over RPC, while another would be to assume the importer is on the same machine as the HTTP API and send the multi-part reader, or the readers to individual parts, over a channel.
+
+As this endpoint will be used for streaming large amounts of data it has the potential to be somewhat difficult to use as system limits could be pushed and errors could occur after annoyingly long time investments.  Some questions that may come up as useability becomes a bigger concern:
+
+* Is it possible that remote users pushing data to the endpoint could overload cluster's resources by pushing at a rate faster than processing can occur in cluster? If this is an issue how can the endpoint maintain connections in a resource-aware non-degrading way?  Will it be worthwhile to open up an endpoint running a different protocol (i.e. libp2p stream-muxer)?
```

Comment:
We might choose to provide the HTTP API via libp2p which might be a good thing for this too.